File: dsa/hw01/code/src/my_ints.py
"""
This code will be used to read files from sample_inputs
then write the numbers in another file making sure no number is repeated
"""
from pathlib import Path

# Defining the base directory as the current directory where the script is running
base_dir = Path(__file__).parent.parent.parent

# Input and output files
input_file = base_dir / "sample_inputs" / "sample_02.txt"
output_file = base_dir / "sample_results" / "results_02.txt"

max_int = 1023
min_int = -1023

# initialise an empty set to ensure uniqueness of the numbers
numbers = set()

# Read the numbers from the input file and add them to the set if they are within the range and not already present
with open(input_file, "r") as f:
    for line in f:
        stripped_line = line.strip()
        if stripped_line:
            try:
                number = int(stripped_line)  # Try converting to int
            except ValueError:
                number = float(stripped_line)
        if max_int >= number >= min_int and number not in numbers:
            numbers.add(number)

# Write the unique numbers to the output file in ascending order
with open(output_file, "w") as f:
    for number in sorted(numbers):
        f.write(f"{number}\n")

File: dsa/hw01/code/src/my_ints.js
const fs = require('fs');
const path = require('path');

// Define the paths
const inputFilePath = path.join(__dirname, '..', 'sample_inputs', 'sample_01.txt');
const outputDirPath = path.join(__dirname, '..', 'sample_results');
const outputFilePath = path.join(outputDirPath, 'results_01.txt');

// Max and min values
const maxInt = 1023;
const minInt = -1013;

// Create a set to hold unique numbers
const numbers = new Set();

// Read the input file
fs.readFile(inputFilePath, 'utf8', (err, data) => {
    if (err) {
        console.error(`Error reading file: ${err}`);
        return;
    }

    const lines = data.split('\n');

    lines.forEach(line => {
        const strippedLine = line.trim();  // Remove whitespace
        if (strippedLine) {  // Check if the line is not empty
            const number = parseFloat(strippedLine);  // Convert to float
            if (!isNaN(number) && number >= minInt && number <= maxInt) {
                numbers.add(number);  // Add to the set if valid
            }
        }
    });

    // Convert the set to an array and sort it
    const sortedNumbers = Array.from(numbers).sort((a, b) => a - b);

    // Ensure the output directory exists
    fs.mkdir(outputDirPath, { recursive: true }, (err) => {
        if (err) {
            console.error(`Error creating directory: ${err}`);
            return;
        }

        // Write the results to the output file
        fs.writeFile(outputFilePath, sortedNumbers.join('\n'), (err) => {
            if (err) {
                console.error(`Error writing file: ${err}`);
                return;
            }
            console.log('Results saved successfully!');
        });
    });
});
